### PR TITLE
Update V3 node

### PR DESCRIPTION
**Pull request by n-h-diaz**
Adds support for property filters, edge filters, and chaining to V3 node. Also fixes the property api to support empty responses.

I ended up adding 6 SQL statements: 
- out arc, single hop
- out arc, chained hops
- in arc, single hop
- in arc, chained hop
- shared subquery for filtering properties
- shared subquery for filtering edges

I split up the single vs chained hops because it was taking too long to return the full response for chained hops. (This comes from processing the edge path, which can be very large and costly.) Instead I only return the name and dcid of the leaf node (and not provenance or types). This is consistent with the current v2 node behavior. I set the limit for chained hops to 10, which can take anywhere from a few hundred milliseconds to ~5 seconds for some of the examples I tested. Adding filters will increase the time too.

Still TODO: support pagination, probably test this more

### CODE REVIEW

**Review by keyurva — COMMENTED** [review]
This is brilliant!

File: internal/server/spanner/query.go [hunk 1]
```diff
@@ -53,38 +58,162 @@ var statements = struct {
...
+	FROM (
+		SELECT
+			*
+		FROM
+			GRAPH_TABLE (
+				DCGRAPH MATCH (m:Node
+				WHERE
+					m.subject_id IN UNNEST(@ids))-[e:Edge
+				WHERE
+					e.predicate = @predicate]->{1,10}(n:Node)
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go]: Done

File: internal/server/spanner/query.go [hunk 2]
```diff
@@ -26,11 +27,15 @@ import (
 
 // SQL / GQL statements executed by the SpannerClient
 var statements = struct {
-	getPropsBySubjectID       string
-	getPropsByObjectID        string
-	getEdgesBySubjectID       string
-	getEdgesByObjectID        string
-	getObsByVariableAndEntity string
+	getPropsBySubjectID        string
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go]: Done

File: internal/server/spanner/dsutil.go [hunk 3]
```diff
@@ -25,19 +25,21 @@ import (
...
-func nodePropsToNodeResponse(props []*Property) *v3.NodeResponse {
+// nodePropsToNodeResponse converts a map from subject id to its properties to a NodeResponse proto.
+func nodePropsToNodeResponse(propsBySubjectID map[string][]*Property) *v3.NodeResponse {
 	nodeResponse := &v3.NodeResponse{
 		Data: make(map[string]*v2.LinkedGraph),
 	}
 
-	for _, prop := range props {
-		linkedGraph, ok := nodeResponse.Data[prop.SubjectID]
+	for subjectID, props := range propsBySubjectID {
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/dsutil.go:43]: Yes I think so! Would it be okay to do as a follow up when we add pagination? It looks like the v2 behavior is to order the response in the request order. Do you have a preference between this vs sorted?

File: internal/server/spanner/query.go [hunk 4]
```diff
@@ -53,38 +58,162 @@ var statements = struct {
...
+			)
+		UNION ALL
+		SELECT
+			*
+		FROM
+			GRAPH_TABLE (
+				DCGraph MATCH -[e:Edge
+				WHERE
+					e.subject_id IN UNNEST(@ids)
+					AND e.object_value IS NOT NULL%[1]s]-> RETURN e.subject_id,
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go]: Done

File: internal/server/spanner/query.go [hunk 5]
```diff
@@ -147,27 +280,66 @@ func (sc *SpannerClient) GetNodeProps(ctx context.Context, ids []string, out boo
...
+		} else {
+			template = fmt.Sprintf(statements.getEdgesByObjectID, filterProps)
 		}
 	}
 
+	// Attach filters.
+	i := 0
+	for p, v := range arc.Filter {
+		template += fmt.Sprintf(statements.filterEdges, i)
+		params["p"+strconv.Itoa(i)] = p
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go]: Done

File: internal/server/spanner/query.go [hunk 6]
```diff
@@ -147,27 +280,66 @@ func (sc *SpannerClient) GetNodeProps(ctx context.Context, ids []string, out boo
...
 	if len(ids) == 0 {
 		return edges, nil
 	}
+	for _, id := range ids {
+		edges[id] = []*Edge{}
+	}
 
-	var stmt spanner.Statement
+	// Validate input.
+	if arc.Decorator != "" && (arc.SingleProp == "" || arc.SingleProp == "*" || len(arc.BracketProps) > 0) {
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go]: Done

File: internal/server/spanner/query.go [hunk 7]
```diff
@@ -53,38 +58,162 @@ var statements = struct {
...
+				WHERE
+					e.predicate = @predicate]->{1,10}(n:Node)
+				WHERE 
+					m != n
+				RETURN DISTINCT m.subject_id,
+					n.subject_id as object_id,
+					'' as object_value,
+					n.name
+			)
+		UNION ALL
```
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go:140]: The performance ranged from a <1s to timing out for large queries (ie >5min). Chaining (especially for in arcs) is slow, and so is adding filters. I think generally this is okay, but could be problematic for ObsCollection. Maybe we could set a lower limit for ObsCollection / allow caller to specify number of hops. I can also try optimizing the queries more

File: internal/server/spanner/dsutil.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (keyurva) [reply internal/server/spanner/dsutil.go:43]: Follow up SG. Should continue with the v2 behavior (request order).

File: internal/server/spanner/query.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (keyurva) [reply internal/server/spanner/query.go:140]: Can we note queries that take longer than a few seconds (let's say max 10 secs)? We either don't want to expose those queries or do mitigations (indexes, pre-computations, etc.) to bring them under the threshold.

File: internal/server/spanner/query.go [hunk 8]
```diff
@@ -309,16 +323,16 @@ func (sc *SpannerClient) GetNodeEdgesByID(ctx context.Context, ids []string, arc
 	var template string
 	switch arc.Out {
 	case true:
-		if arc.Decorator == "+" {
-			template = statements.getChainedEdgesBySubjectID
+		if arc.Decorator == CHAIN {
+			template = fmt.Sprintf(statements.getChainedEdgesBySubjectID, MAX_HOPS)
```
Comment (keyurva) [inline internal/server/spanner/query.go:327]: nit: Consider formatting the string with constant args (like `MAX_HOPS`) when defining the statement itself.

File: internal/server/spanner/dsutil.go [hunk 9]
```diff
@@ -25,6 +25,15 @@ import (
 	"google.golang.org/protobuf/proto"
 )
 
+const (
+	// Indicates that all properties should be returned.
+	WILDCARD = "*"
+	// Indicates that
+	CHAIN = "+"
+	// Maximum number of edge hops to traverse for chained properties.
+	MAX_HOPS = 10
```
Comment (keyurva) [inline internal/server/spanner/dsutil.go:34]: nit: Is max hops used in other places or only when making spanner queries? If so, consider moving it to `query.go`.

**Review by keyurva — APPROVED** [review]

File: internal/server/spanner/query.go (hunk 8 above)
Comment (n-h-diaz) [reply internal/server/spanner/query.go]: Done

File: internal/server/spanner/dsutil.go (hunk 9 above)
Comment (n-h-diaz) [reply internal/server/spanner/dsutil.go]: Done

File: internal/server/spanner/query.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (n-h-diaz) [reply internal/server/spanner/query.go:145]: I'll add these to the notes doc! I think I'm a bit worried for ObsCollection case, maybe we can follow up with the Spanner team to see if they have suggestions for how to improve our schema/queries

**Comment by n-h-diaz** [comment]
Thanks for the review!

**Pull request merged**